lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp

```diff
@@ -700,6 +727,59 @@ SILValue BlockState::reduceValuesAtEndOfBlock(RLEContext &Ctx, LSLocation &L) {
   return TheForwardingValue;
 }
 
+SILValue
+BlockState::createCopyOfForwardingValueIfNeeded(SILValue TheForwardingValue,
+                                                RLEContext &Ctx) {
+  auto *Fn = TheForwardingValue->getFunction();
+  if (TheForwardingValue->getType().isTrivial(*Fn))
+    return TheForwardingValue;
+
+  auto &ForwardedValues = Ctx.getForwardedValues();
+  auto &NewCopies = Ctx.getNewCopies();
+  auto &NewPhiArgs = Ctx.getNewPhiArgs();
+
+  auto createCopyAndSetState =
+      [&NewCopies](SILValue TheForwardingValue) -> SILValue {
+    auto Copy =
+        SILBuilderWithScope(getInsertAfterPoint(TheForwardingValue).getValue())
+            .emitCopyValueOperation(TheForwardingValue.getLoc(),
+                                    TheForwardingValue);
+    NewCopies.insert(cast<CopyValueInst>(Copy));
+    return Copy;
+  };
+
+  if (isa<CopyValueInst>(TheForwardingValue)) {
```

Comment:
This is a "pro-actively" created copy_value from processWrite or processRead, right?
Can you mention this in a comment here?

Wouldn't it be easier to not "pro-actively" create the copy_values, but just create all (including the first one) here?
Also, is it always sufficient to create the copy_value at the location of the forwarding value? What if the load to forward is outside a loop and the load to replace inside a loop? In this case a copy_value must be inserted in the loop.

Reply:
Yes, it is the proactively created copy_value from processWrite and processRead. This can also include copy_value of struct_extract if the redundant load was a field load and the available dominating load was a load of its struct.

We can't insert the first copy_value here because we need it to be available before RLEContext::collectLocationValues. This is because it ends up calling SSAUpdater::getValueInMiddleOfBlock which can insert a phi arg of the loaded value which ends its lifetime in OSSA. And if we have any post dominating use of the loaded value, it will end up being a value after a consuming use. `rle_needscopy1` in `redundant_load_elim_ossa_complex.sil` shows this case

Reply:
Thanks for pointing out if its sufficient to create a copy_value at the location of the forwarding value. I realized RLE on OSSA is not removing such a load. Added test` rle_redundantload_does_not_postdominate2` and `rle_redundantload_does_not_postdominate3` in `redundant_load_elim_ossa_complex.sil`. I'll investigate this further.

Reply:
I see, thanks

Reply:
I updated the PR with the fix and tests. I am now using SILLoopAnalysis to determine if we need a copy at the redundant load.

Reply:
SILLoopAnalysis is not the right tool for this. There might be cycles in the CFG which are not natural loops.

Reply:
Thanks for suggesting ValueLifetimeAnalysis as an alternative offline.